Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift

```diff
@@ -43,250 +43,183 @@ extension BinaryInteger {
-
-        // It's easier to just special-case zero than handle it in the main algorithm.
-        guard .zero != self else {
-            return [UInt8(ascii: "0")]
-        }
-
-        let maximumDigits = Self.maximumDecimalDigitsForUnsigned(bitWidth: self.magnitudeBitWidth)
-        var actualDigits: Int = Int.min // Actually initialised inside the closure below, but the compiler mistakenly demands a default value anyway.
-
-        return ContiguousArray(unsafeUninitializedCapacity: maximumDigits) { buffer, initialisedCount in
-            actualDigits = numericStringRepresentation(intoEndOfBuffer: &buffer[...])
-
-            let unusedDigits = maximumDigits - actualDigits
-            assert(0 <= unusedDigits, "Negative unused digits \(unusedDigits)!  Expected at most \(maximumDigits) digit(s) but emitted \(actualDigits).")
-
-            if 0 < unusedDigits {
-                buffer[0..<unusedDigits].initialize(repeating: 0) // The buffer is permitted to be partially uninitialised, but only at the end.  So we have to initialise the unused portion at the start, in principle.  Technically this probably doesn't matter given we never subsequently read this part of the buffer, but there's no way to express that such that the compiler can ensure it for us.
-            }
-
-            initialisedCount = maximumDigits
-        }[(maximumDigits - actualDigits)...]
+/// Returns the largest `exponent` and `power` in `pow(10, exponent) <= UInt.max + 1`.
+///
+/// The `exponent` is also the maximum number of decimal digits needed to represent a binary integer
+/// in the range of `0 ..< power`. Another method is used to estimate the total number of digits, however.
+/// This is so that binary integers can be rabased and encoded in the same loop.
+///
+/// ```
+/// 32-bit: (exponent:  9, power:           1000000000)
+/// 64-bit: (exponent: 19, power: 10000000000000000000)
+/// ```
+///
+private func maxDecimalExponentAndPowerForUnsignedIntegerWord() -> (exponent: Int, power: UInt) {
+    var exponent = 1 as  Int
+    var power = 0010 as UInt
+    
+    while true {
+        let next = power.multipliedReportingOverflow(by: 10)
+        if  next.overflow { break }
+        
+        exponent += 1
+        power = next.partialValue
     }
+    
+    return (exponent: exponent, power: power)
+}
+
+/// Returns an upper bound for the [number of decimal digits][algorithm] needed
+/// to represent an unsigned integer with the given `bitWidth`.
+///
+/// [algorithm]: https://www.exploringbinary.com/number-of-decimal-digits-in-a-binary-integer
+///
+/// - Parameter bitWidth: The unsigned binary integer's bit width.
+///
+private func maxDecimalDigitCountForUnsignedInteger(bitWidth: Int) -> Int {
+    // - Int.init(some BinaryFloatingPoint) rounds to zero.
+    // - Double.init(exactly:) and UInt.init(_:) for correctness.
+    // - log10(2.0) is: 1.0011010001000001001101⌈01...⌉ * 2^(-2).
+    return Int(Double(exactly: UInt(bitWidth))! * log10(2.0)) + 1
```

Comment:
This doesn't work correctly for `bitWidth`s of 1 or less.  Those aren't values that will appear at runtime, but it might be worth putting an assert here to help ensure that just in case some future refactor exposes this to wider use.

And/or document this limitation in the function documentation.

Also, changing the parameter type to `UInt` will at least preclude negative values.

Reply:
I apologise, you're quite right.  I misunderstood the algorithm here (I think re. the rounding behaviour).  I've now tested it empirically and indeed it works as intended.

The value of 1 for bit width 0 is okay - that case shouldn't ever arise in practice anyway - although a `precondition(0 < bitWidth, …)` might be wise (or an assert).

Reply:
The zero bit width case occurs when an integer's `words` collection is empty. I don't have an integer like that, but you could perhaps justify an unsigned big integer normalizing zero to `[]`. In any case, I have unit tests for empty collections, which result in `"0"`.